app/controllers/application_controller.rb

```diff
@@ -5,6 +5,140 @@ class ApplicationController < Sinatra::Base
+      session[:user_id] = user.id
+      redirect '/tweets', locals: {message: "Sign up successful, please log in now."}
+    else
+      redirect '/signup', locals: {message: "Sign up failed, try again."}
+    end
+  end
+
+  get "/login" do
+    if !logged_in?
+      erb :'/users/login'
+    else
+      redirect '/tweets', locals: {message: "You're already logged in!"}
+    end
+  end
+
+  post "/login" do
+    user = User.find_by(username: params[:username])
+    if user && user.authenticate(params[:password])
+      session[:user_id] = user.id
+      redirect "/tweets", locals: {message: "Successfully logged in."}
+    else
+      redirect "/login", locals: {message: "Log in failed, try again."}
+    end
+  end
+
+  get "/logout" do
+    if logged_in?
+      session.clear
+      redirect "/login", locals: {message: "Successfully logged out."}
+    else
+      redirect "/login", locals: {message: "Can't log out, you're not logged in."}
+    end
+  end
+
+  get "/users/:slug" do
+    @user = User.find_by_slug(params[:slug])
+    erb :'/users/show'
+  end
+
+  get "/tweets" do
+    if logged_in?
+      @tweets = Tweet.all
+      erb :'/tweets/tweets'
+    else
+      redirect "/login", locals: {message: "Please log in to see that."}
+    end
+  end
+
+  get "/tweets/new" do
+    if logged_in?
+      erb :'/tweets/new'
+    else
+      redirect "/login", locals: {message: "Please log in to see that."}
+    end
+  end
+
+  post "/tweets" do
+    if logged_in? && params[:content] != ""
+      Tweet.create(content: params[:content], user_id: session[:user_id])
+    elsif params[:content] == ""
+      redirect "/tweets/new", locals: {message: "Tweets can't be empty."}
+    else
+      redirect "/login", locals: {message: "Please log in to see that."}
+    end
+  end
+
+  get "/tweets/:id" do
+    if logged_in?
+      @tweet = Tweet.find(params[:id])
+      erb :'/tweets/show'
+    else
+      redirect "/login", locals: {message: "Please log in to see that."}
+    end
+  end
+
+  get "/tweets/:id/edit" do
+    if logged_in?
+      @tweet = Tweet.find(params[:id])
+      erb :'/tweets/edit'
+    else
+      redirect "/login", locals: {message: "Please log in to see that."}
+    end
+  end
+
+  post "/tweets/:id" do
+    @tweet = Tweet.find(params[:id])
+    if logged_in? && params[:content] != ""
+      @tweet.update(content: params[:content])
+      redirect "/tweets/#{@tweet.id}"
+    elsif params[:content] == ""
+      redirect "/tweets/#{@tweet.id}/edit", locals: {message: "Tweets can't be empty."}
+    else
+      redirect "/login", locals: {message: "Please log in to see that."}
+    end
+  end
+
+  delete "/tweets/:id/delete" do
+    @tweet = Tweet.find(params[:id])
+    if logged_in? && @tweet.user == current_user
+      @tweet.destroy
+      redirect "/tweets"
+    else
+      redirect "/login", locals: {message: "Please log in to see that."}
+    end
+  end
+
+
+  helpers do
+    def current_user
+      User.find(session[:user_id])
```

Comment:
Because in the lifespan of a request, you might call `current_user` multiple times, we'd prefer this method to only hit the database once and then 'cache' the user instance into an instance variable that could be returned on any subsequent call to `current_user`.

```ruby
def current_user
  @current_user ||= User.find(session[:id) if session[:id]
end
```

This pattern of conditionally setting an instance variable or returning it if it's already defined using the `||=` (or-equals) operator is called Memoization. ([read more](http://gavinmiller.io/2013/basics-of-ruby-memoization/) and [here](http://www.justinweiss.com/articles/4-simple-memoization-patterns-in-ruby-and-one-gem/))